packages/itwinui-react/src/utils/components/OverflowContainer.tsx

```diff
@@ -96,6 +104,224 @@ if (process.env.NODE_ENV === 'development') {
   OverflowContainerContext.displayName = 'OverflowContainerContext';
 }
 
+/**
+ * Hook that returns the number of items that should be visible based on the size of the container element.
+ *
+ * The returned number should be used to render the element with fewer items.
+ *
+ * @private
+ * @param items Items that this element contains.
+ * @param dimension 'horizontal' (default) or 'vertical'
+ * @returns [callback ref to set on container, stateful count of visible items]
+ *
+ * @example
+ * const items = Array(10).fill().map((_, i) => <span>Item {i}</span>);
+ * const [ref, visibleCount] = useOverflow(items);
+ * ...
+ * return (
+ *   <div ref={ref}>
+ *     {items.slice(0, visibleCount)}
+ *   </div>
+ * );
+ */
+const useOverflow = <T extends HTMLElement>(
+  itemsCount: number,
+  orientation: 'horizontal' | 'vertical' = 'horizontal',
+) => {
+  /** `[number, number]` means that we're still guessing. `null` means that we got the correct `visibleCount`. */
+  type GuessRange = [number, number] | null;
+
+  /** First guess of the number of items that overflows. We refine this guess with subsequent renders. */
+  const STARTING_MAX_ITEMS_COUNT = 32;
+
+  const containerRef = React.useRef<T>(null);
+
+  const initialVisibleCount = React.useMemo(
+    () => Math.min(itemsCount, STARTING_MAX_ITEMS_COUNT),
+    [itemsCount],
+  );
+  const initialVisibleCountGuessRange = React.useMemo(
+    () => [0, initialVisibleCount] satisfies GuessRange,
+    [initialVisibleCount],
+  );
+
+  const [visibleCount, _setVisibleCount] =
+    React.useState<number>(initialVisibleCount);
+  const [visibleCountGuessRange, setVisibleCountGuessRange] =
+    React.useState<GuessRange>(initialVisibleCountGuessRange);
+
+  const isStabilized = visibleCountGuessRange == null;
+
+  /**
+   * Ensures that `visibleCount <= itemsCount`
+   */
+  const setVisibleCount = React.useCallback(
+    (setStateAction: React.SetStateAction<typeof visibleCount>) => {
+      _setVisibleCount((prev) => {
+        const newVisibleCount =
+          typeof setStateAction === 'function'
+            ? setStateAction(prev)
+            : setStateAction;
+
+        return Math.min(newVisibleCount, itemsCount);
+      });
+    },
+    [itemsCount],
+  );
+
+  const [resizeRef] = useResizeObserver(
+    React.useCallback(() => {
+      setVisibleCount(initialVisibleCount);
+      setVisibleCountGuessRange(initialVisibleCountGuessRange);
+    }, [initialVisibleCount, initialVisibleCountGuessRange, setVisibleCount]),
+  );
```

Comment:
To keep the overflow code centralized and reduce repetition, you might need a wrapper component in `OverflowContainer.tsx`.

The nested hierarchy would be something like:

```
- Breadcrumbs/ButtonGroup/etc
- OverflowContainerWrapper
  - useContainerWidth
  - OverflowContainerMain (key)
    - useOverflow
```

Reply:
There are two issues in that commit: `keyProp` and `refProp`. We cannot destructure these things because `key` and `ref` are not regular props; they are reserved get removed from the props object by React.

I don't see proper docs for this; the closest thing I found was the [representation of a JSX node](https://react.dev/reference/react/createElement#returns), which lists out `key` and `ref` separately from `props`. You can also verify the behavior at runtime; I believe `keyProp` and `refProp` will always be `undefined`.

Reply:
Another issue I've noticed is that in multi-select ComboBox, if you select enough items then make the window narrower, the tag container correctly collapses based on available width, but then when you resize it back, it doesn't expand back. Maybe this is ok, because it's a rare scenario?

To handle this correctly, we would need to move the ResizeObserver back into the ComboBox.